modules/snooker/src/main/ts/ephox/snooker/operate/TransformOperations.ts

```diff
@@ -1,79 +1,165 @@
-import { Arr } from '@ephox/katamari';
-import { SugarElement } from '@ephox/sugar';
+import { Arr, Fun, Optional } from '@ephox/katamari';
+import { Attribute, SugarElement, SugarNode } from '@ephox/sugar';
 
 import * as Structs from '../api/Structs';
 import { TableSection } from '../api/TableSection';
+import { isHeaderCell, isHeaderCells } from '../lookup/Type';
 import * as GridRow from '../model/GridRow';
+import * as CellUtils from '../util/CellUtils';
 
 type CompElm = (e1: SugarElement, e2: SugarElement) => boolean;
 type Subst = (element: SugarElement, comparator: CompElm) => SugarElement;
 type CellReplacer = (cell: Structs.ElementNew, comparator: CompElm, substitute: Subst) => Structs.ElementNew;
+type ScopeGenerator = (cell: Structs.ElementNew, rowIndex: number, colIndex: number) => Optional<null | string>;
+type ReplacePredicate = (cell: Structs.ElementNew, rowIndex: number, colIndex: number) => boolean;
+
+const notStartRow = (grid: Structs.RowCells[], rowIndex: number, colIndex: number, comparator: CompElm): boolean =>
+  GridRow.getCellElement(grid[rowIndex], colIndex) !== undefined && (rowIndex > 0 && comparator(GridRow.getCellElement(grid[rowIndex - 1], colIndex), GridRow.getCellElement(grid[rowIndex], colIndex)));
+
+const notStartColumn = (row: Structs.RowCells, index: number, comparator: CompElm): boolean =>
+  index > 0 && comparator(GridRow.getCellElement(row, index - 1), GridRow.getCellElement(row, index));
+
+const alreadyProcessed = (grid: Structs.RowCells[], rowIndex: number, colIndex: number, comparator: CompElm): boolean =>
+  notStartRow(grid, rowIndex, colIndex, comparator) || notStartColumn(grid[rowIndex], colIndex, comparator);
+
+const rowReplacerPredicate = (targetRow: Structs.RowCells, columnHeaders: boolean[]): ReplacePredicate => {
+  const entireTableIsHeader = Arr.forall(columnHeaders, Fun.identity) && isHeaderCells(targetRow.cells);
+  return entireTableIsHeader ? Fun.always : (cell, _rowIndex, colIndex) => {
+    const type = SugarNode.name(cell.element);
+    return !(type === 'th' && columnHeaders[colIndex]);
+  };
+};
+
+const columnReplacePredicate = (targetColumn: Structs.ElementNew[], rowHeaders: boolean[]): ReplacePredicate => {
+  const entireTableIsHeader = Arr.forall(rowHeaders, Fun.identity) && isHeaderCells(targetColumn);
+  return entireTableIsHeader ? Fun.always : (cell, rowIndex, _colIndex) => {
+    const type = SugarNode.name(cell.element);
+    return !(type === 'th' && rowHeaders[rowIndex]);
+  };
+};
+
+const determineScope = (applyScope: boolean, element: SugarElement<HTMLTableCellElement>, newScope: 'row' | 'col', isInHeader: boolean) => {
```

Comment:
We don't in many other cases so why would we want it for this single function? It's only exported functions that we require it to be added normally.

Reply:
I just thought it might be good since the function can return `string | null`. Although, it is clear from the function contents what it is doing so am fine with not having a return type.